sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java

```diff
@@ -0,0 +1,510 @@
+   *
+   * <p>This should only be called with an ended {@code Span}.
+   *
+   * @return an immutable representation of all the data from this {@code Span}.
+   * @throws IllegalStateException if the Span doesn't have RECORD_EVENTS option.
+   */
+  public SpanData toSpanData() {
+    synchronized (this) {
+      Preconditions.checkState(
+          hasBeenEnded, "Calling toSpanData() on a running span is not supported.");
+      return SpanData.create(
+          getContext(),
+          parentSpanId,
+          resource,
+          name,
+          kind,
+          toSpanDataTimestamp(timestampConverter, startNanoTime),
+          attributes == null
+              ? Collections.<String, AttributeValue>emptyMap()
+              : Collections.unmodifiableMap(attributes),
+          events == null
+              ? Collections.<SpanData.TimedEvent>emptyList()
+              : toSpanDataTimedEvents(events, timestampConverter),
+          links == null
+              ? Collections.<Link>emptyList()
+              : Collections.unmodifiableList(new ArrayList<Link>(links)),
+          getStatusWithDefault(),
+          toSpanDataTimestamp(timestampConverter, endNanoTime));
+    }
+  }
+
+  @Override
+  public void setAttribute(String key, String value) {
+    setAttribute(key, AttributeValue.stringAttributeValue(value));
+  }
+
+  @Override
+  public void setAttribute(String key, long value) {
+    setAttribute(key, AttributeValue.longAttributeValue(value));
+  }
+
+  @Override
+  public void setAttribute(String key, double value) {
+    setAttribute(key, AttributeValue.doubleAttributeValue(value));
+  }
+
+  @Override
+  public void setAttribute(String key, boolean value) {
+    setAttribute(key, AttributeValue.booleanAttributeValue(value));
+  }
+
+  @Override
+  public void setAttribute(String key, AttributeValue value) {
+    Preconditions.checkNotNull(key, "key");
+    Preconditions.checkNotNull(value, "value");
+    synchronized (this) {
+      if (hasBeenEnded) {
+        logger.log(Level.FINE, "Calling setAttribute() on an ended Span.");
+        return;
+      }
+      getInitializedAttributes().putAttribute(key, value);
+    }
+  }
+
+  @Override
+  public void addEvent(String name) {
+    addTimedEvent(TimedEvent.create(clock.nowNanos(), name));
+  }
+
+  @Override
+  public void addEvent(String name, Map<String, AttributeValue> attributes) {
+    addTimedEvent(TimedEvent.create(clock.nowNanos(), name, attributes));
+  }
+
+  @Override
+  public void addEvent(Event event) {
+    addTimedEvent(TimedEvent.create(clock.nowNanos(), event));
+  }
+
+  private void addTimedEvent(TimedEvent timedEvent) {
+    synchronized (this) {
+      if (hasBeenEnded) {
+        logger.log(Level.FINE, "Calling addEvent() on an ended Span.");
+        return;
+      }
+      getInitializedEvents().add(timedEvent);
+      totalRecordedEvents++;
+    }
+  }
+
+  private static List<SpanData.TimedEvent> toSpanDataTimedEvents(
```

Comment:
why do we have all the to SpanData methods?

Reply:
Removed.